### ISSUE TITLE
[BUG] VM can’t boot due to filesystem corruption

### ISSUE THREAD

**Issue by guangbochen** [harvester/harvester#2906]
**Describe the bug**
When running VMs on harvester, rarely the VMs could not boot up due to filesystem corruption.

**To Reproduce**
1. Start VM
2. Run IO on VM
3. delete any one of the replicas and stop VM (parallel operations)
4. loop from step 1

If we meet the above issue, Start VM would be stuck.

**Expected behavior**
Works normally. Turn on/off VM would be OK. 

**Support bundle**

**Environment**
 - Harvester ISO version: v1.0.3

**Additional context**
This happens when volume rebuilding is involved, for example, turning on/off VMs and powering cycle hosts.
It seems related to https://github.com/longhorn/longhorn/issues/4354

**Comment by Martin-Weiss** [comment]
@guangbochen  - what type of filesystem do you run within your VM? Can you add details?
I believe if you use a filesystem with 4k I/O size instead of 512 bytes you might not be able to reproduce..

**Comment by Vicente-Cheng** [comment]
We running with `OpenSUSE-Leap-15.3`, the root filesystem is XFS with 512 sector size.
We could reproduce the corruption issue in the above environment with harvester v1.0.3.

**Comment by Martin-Weiss** [comment]
> We running with `OpenSUSE-Leap-15.3`, the root filesystem is XFS with 512 sector size. We could reproduce the corruption issue in the above environment with harvester v1.0.3.

Ok - then it is pretty likely https://github.com/longhorn/longhorn/issues/4354 and the current workaround should be to use a different filesystem layout using 4k instead of 512.

**Comment by Vicente-Cheng** [comment]
For v1.0.3, the current workaround is to re-generate the cloud image to use a filesystem with 4k sector size for metadata/data/log.

We could see the related fix ([[BUG] data corruption due to COW and block size not being aligned during rebuilding replicas](https://github.com/longhorn/longhorn/issues/4354#top)) is involved in Longhorn v1.3.2. So we could verify this with LH v1.3.2

**Comment by Vicente-Cheng** [comment]
Update with test result on my local environment, QA would do a whole test including the regression.

environment rootfs:
```
vm001:~ # xfs_info /dev/vda3
meta-data=/dev/vda3              isize=512    agcount=4, agsize=653887 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=0, rmapbt=0
         =                       reflink=0
data     =                       bsize=4096   blocks=2615547, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
```

tested with the following steps:

1. start VM
2. run 4k random write IO / Delete random replica  (these two operations are parallel)
3. wait 20-60 seconds then stop VM
4. loop back to 1.

pass 626 rounds and do not meet any corruption issues.

**Comment by TachunLin** [comment]
Verified fixed on Harvester `master-5ee476dd-head` (10/12) with `Longhorn v1.3.2-rc2` merged. 

### Test Result
* Pass 586 rounds of the I/O write test, did not encounter data corruption issue and VM is alive
```
opensuse:~ # xfs_info /dev/vda3
meta-data=/dev/vda3              isize=512    agcount=13, agsize=653887 blks
         =                       sectsz=512   attr=2, projid32bit=1
         =                       crc=1        finobt=1, sparse=0, rmapbt=0
         =                       reflink=0
data     =                       bsize=4096   blocks=7858427, imaxpct=25
         =                       sunit=0      swidth=0 blks
naming   =version 2              bsize=4096   ascii-ci=0, ftype=1
log      =internal log           bsize=4096   blocks=2560, version=2
         =                       sectsz=512   sunit=0 blks, lazy-count=1
realtime =none                   extsz=4096   blocks=0, rtextents=0
```

### Test Information
* Test Environment: 3 nodes harvester on local kvm machine
* Harvester version: `master-5ee476dd-head` (10/12)


### Verification steps
Execute the python script provide by team member to execute start and stop VM
which will execute fio command to run I/O test 
And use another shell script to delete replica paralleled with I/O write test. 
  
1. Start VM
2. run 4k random write IO / Delete random replica (these two operations are parallel)
3. wait 20-60 seconds then stop VM
4. loop back to 1.

**Comment by TachunLin** [comment]
#### Pass the following volume operation regression test: 
1. Create volume (Harvester, Longhorn storage class)
2. Create volume from image
3. Unmount volume from VM
4. Delete volume in use and not in use
5. Export volume to image 
6. Create VM from the exported image
7. Edit volume to increase size 
8. Delete volume in use 
9. Clone volume 
10. Take volume snapshot
11. Restore volume snapshot

**Comment by TachunLin** [comment]
Verified fixed on Harvester master-5ee476dd-head (10/12) with Longhorn v1.3.2-rc2 on data corruption. 
And also pass the volume related regression test as listed above. Close this issue.
